examples/crd_derive_schema.rs

```diff
@@ -0,0 +1,275 @@
+                                          "required_field": {
+                                              "type": "string"
+                                          },
+                                          "with_default": {
+                                              "default": "default_value",
+                                              "type": "string"
+                                          },
+                                        },
+                                        "required": [
+                                            "required_field"
+                                        ],
+                                        "type": "object"
+                                    }
+                                },
+                                "required": [
+                                    "spec"
+                                ],
+                                "title": "Foo",
+                                "type": "object"
+                            }
+                        },
+                        "subresources": {},
+                    }
+                ]
+            }
+        });
+        let expected = serde_json::from_value(output).unwrap();
+        // println!("{}", serde_yaml::to_string(&crd).unwrap());
+        // println!("{}", serde_yaml::to_string(&expected).unwrap());
+        assert_eq!(crd, expected);
+    }
+
+    // All tests are in this test case to avoid multiple test cases from creating/deleting Foo.
+    #[tokio::test]
+    async fn test_create() {
+        // Creating CRD v1 works as expected.
+        let client = Client::try_default().await.unwrap();
+        delete_crd(client.clone()).await.unwrap();
+        assert!(create_crd(client.clone()).await.is_ok());
+
+        // Test creating Foo resource.
+        let namespace = std::env::var("NAMESPACE").unwrap_or("default".into());
+        let foos = Api::<Foo>::namespaced(client.clone(), &namespace);
+        let bar = Foo::new(
+            "bar",
+            FooSpec {
+                required_field: "required".into(),
+                with_default: "".into(),
+                nullable: None,
+                nullable_with_default: None,
+                nullable_with_default_no_skip: None,
+            },
+        );
+        let bar = foos.create(&PostParams::default(), &bar).await.unwrap();
+        // Defaulting happens if None is skipped.
+        assert_eq!(bar.spec.nullable_with_default, Some("default_nullable".into()));
+        // If `null` is sent, defaulting does not happen.
+        // Deserialization does not apply the default either.
+        assert_eq!(bar.spec.nullable_with_default_no_skip, None);
+        // `nullable` field should not exist on the object, but it's set to `None` when deserializing.
+        // TODO Any convenient way to check the raw response?
+        assert_eq!(bar.spec.nullable, None);
+
+        // TODO Any convenient way to send invalid/raw data?
```

Comment:
@clux Is there a way to send/receive untyped data conveniently?

Reply:
Sending untyped, yes. Use `serde_json::json!` and turn that to a vec before passing to [Api::patch](https://docs.rs/kube/0.43.0/kube/struct.Api.html#method.patch).

If you need both, then untyped is harder, you'll need to use [DynamicResource](https://docs.rs/kube/0.43.0/kube/api/struct.Resource.html#method.dynamic) to get a `Resource`, then use `resource::get` (say) with your `Client` directly - easiest to look at [typed.rs](https://github.com/clux/kube-rs/blob/master/kube/src/api/typed.rs#L80-L83) for how this works. It's a little more verbose, becaues you have to specify the output type (can be `serde_json::Value`).

Reply:
At that point though, aren't you just testing that arbitrary json is deserialized correctly by serde? Deserialization isn't super interesting on its own.

Testing strategies wise; I'd say it's more important to test that any type we can put inside a struct comes back through kubernetes in a form we expect (and is coercible into a K), given common serde + schema annotations, rather than what happens if kubernetes returns bad json. If we can cover a reasonable set of cases, we'll run this test case on 1.19 kind.